src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +279,159 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="state">The state associated with the target history entry.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation. If the result is <see langword="true"/>, the navigation should continue.</returns>
+    protected async ValueTask<bool> NotifyLocationChangingAsync(string uri, string? state, bool isNavigationIntercepted)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        var handlerCount = _locationChangingHandlers.Count;
+
+        if (handlerCount == 0)
+        {
+            return true;
+        }
+
+        var cts = new CancellationTokenSource();
+
+        _locationChangingCts = cts;
+
+        var cancellationToken = cts.Token;
+        var context = new LocationChangingContext(uri, state, isNavigationIntercepted, cancellationToken);
+
+        try
+        {
+            if (handlerCount == 1)
+            {
+                var handlerTask = _locationChangingHandlers[0](context);
+
+                if (context.DidPreventNavigation)
+                {
+                    return false;
+                }
+
+                if (!handlerTask.IsCompletedSuccessfully)
+                {
+                    await handlerTask.AsTask().WaitAsync(cancellationToken);
+                }
+            }
+            else
+            {
+                var locationChangingHandlersCopy = ArrayPool<Func<LocationChangingContext, ValueTask>>.Shared.Rent(handlerCount);
+
+                try
+                {
+                    _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+                    var locationChangingTasks = new HashSet<Task>();
+
+                    for (var i = 0; i < handlerCount; i++)
+                    {
+                        var handlerTask = locationChangingHandlersCopy[i](context);
+
+                        if (context.DidPreventNavigation)
+                        {
+                            return false;
+                        }
+
+                        locationChangingTasks.Add(handlerTask.AsTask());
+                    }
+
+                    while (locationChangingTasks.Count != 0)
+                    {
+                        var completedHandler = await Task.WhenAny(locationChangingTasks).WaitAsync(cancellationToken);
+
+                        if (completedHandler.Exception is { } exception)
+                        {
+                            throw exception;
```

Comment:
```suggestion
                            await completedHandler;
```

`throw task.exception` can have different semantics (I believe it overrides the StackTrace and maybe other properties). That is why in some places of the codebase we use [`ExceptionDispatchInfo.Capture(exception).Throw()`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?view=net-6.0#remarks) to preserve the original information.

If you simply await the task, the async machinery takes care of making sure that everything gets properly propagated.